vertical-pod-autoscaler/pkg/utils/vpa/capping.go

```diff
@@ -184,22 +184,39 @@ func applyVPAPolicy(recommendation apiv1.ResourceList, policy *vpa_types.Contain
 
 func applyVPAPolicyForContainer(containerName string,
 	containerRecommendation *vpa_types.RecommendedContainerResources,
-	policy *vpa_types.PodResourcePolicy) (*vpa_types.RecommendedContainerResources, error) {
+	policy *vpa_types.PodResourcePolicy,
+	globalMaxAllowed apiv1.ResourceList) (*vpa_types.RecommendedContainerResources, error) {
 	if containerRecommendation == nil {
 		return nil, fmt.Errorf("no recommendation available for container name %v", containerName)
 	}
 	cappedRecommendations := containerRecommendation.DeepCopy()
-	// containerPolicy can be nil (user does not have to configure it).
 	containerPolicy := GetContainerResourcePolicy(containerName, policy)
-	if containerPolicy == nil {
-		return cappedRecommendations, nil
-	}
 
 	process := func(recommendation apiv1.ResourceList) {
 		for resourceName, recommended := range recommendation {
-			cappedToMin, _ := maybeCapToPolicyMin(recommended, resourceName, containerPolicy)
-			recommendation[resourceName] = cappedToMin
-			cappedToMax, _ := maybeCapToPolicyMax(cappedToMin, resourceName, containerPolicy)
+			var maxAllowed apiv1.ResourceList
+			// containerPolicy can be nil (user does not have to configure it).
+			if containerPolicy != nil {
+				cappedToMin, _ := maybeCapToPolicyMin(recommended, resourceName, containerPolicy)
+				recommendation[resourceName] = cappedToMin
+
+				maxAllowed = containerPolicy.MaxAllowed
+			}
+
+			if globalMaxAllowed != nil {
+				if maxAllowed == nil {
+					maxAllowed = globalMaxAllowed
+				} else {
+					// Set resources from the global maxAllowed if the VPA maxAllowed is missing them.
+					for resourceName, quantity := range globalMaxAllowed {
+						if _, ok := maxAllowed[resourceName]; !ok {
```

Comment:
Let's add a comment here that we only override this if the user did not explicitly set a maximum in their container policy in the VPA.

Reply:
There is already https://github.com/kubernetes/autoscaler/blob/04e33401fe93d83320414e87f30570cdc24586de/vertical-pod-autoscaler/pkg/utils/vpa/capping.go#L210 which is more or less stating the same thing.
Let me know if you have suggestions for improvements.